### PR TITLE
fix(obstacle_avoidance_planner, interpolation): improve slerp and fix yaw

### LINKED ISSUE #966: Yaw calculation is unstable in obstacle_avoidance_planner when skipping optimization.
### Checklist

- [X] I've read the [contribution guidelines](https://github.com/autowarefoundation/autoware/blob/main/CONTRIBUTING.md).
- [X] I've searched other issues and no duplicate issues were found.
- [X] I'm convinced that this is not my fault but a bug.

### Description

We can skip optimization in obstacle_avoidance_planner (then, just resample path from behavior module).
In this condition, yaw calculation is unstable and this causes some failure in other modules (e.g. detection area in obstacle stop planner is expanded, and it detects obstacles outside the trajectory, then do stop planning).

### Expected behavior

Yaw calculation is stable

### Actual behavior

Yaw calculation is unstable

### Steps to reproduce

1. fix yaw calculation algorithm

### Versions

_No response_

### Possible causes

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by angry-crab**
Signed-off-by: Xinyu Wang <xinyu.wang@tier4.jp>

## Description

#1499 Fix yaw interpolation and improve performance.
Since the current yaw interpolation is not stable, I just switched to the implementation of `fillYawInTrajectoryPoint`.
This PR also somehow fixes #966 by correcting yaw afterwards.

## Before

```
[component_container_mt-15] [INFO] [1661409128.691758302] [obstacle_avoidance_planner.time]:     generateFineTrajectoryPoints:= 3.928 [ms]
[component_container_mt-15] [INFO] [1661409128.756615001] [obstacle_avoidance_planner.time]:     generateFineTrajectoryPoints:= 2.922 [ms]
[component_container_mt-15] [INFO] [1661409128.865204451] [obstacle_avoidance_planner.time]:     generateFineTrajectoryPoints:= 5.034 [ms]
[component_container_mt-15] [INFO] [1661409128.964733989] [obstacle_avoidance_planner.time]:     generateFineTrajectoryPoints:= 5.834 [ms]
[component_container_mt-15] [INFO] [1661409129.110849547] [obstacle_avoidance_planner.time]:     generateFineTrajectoryPoints:= 4.85 [ms]
[component_container_mt-15] [INFO] [1661409129.163539277] [obstacle_avoidance_planner.time]:     generateFineTrajectoryPoints:= 3.794 [ms]
[component_container_mt-15] [INFO] [1661409129.501548288] [obstacle_avoidance_planner.time]:     generateFineTrajectoryPoints:= 25.244 [ms]
[component_container_mt-15] [INFO] [1661409129.527538449] [obstacle_avoidance_planner.time]:     generateFineTrajectoryPoints:= 22.677 [ms]
[component_container_mt-15] [INFO] [1661409129.552173786] [obstacle_avoidance_planner.time]:     generateFineTrajectoryPoints:= 7.98 [ms]
```
app

## After

```
[component_container_mt-15] [INFO] [1661412920.206313216] [obstacle_avoidance_planner.time]:     generateFineTrajectoryPoints:= 0.227 [ms]
[component_container_mt-15] [INFO] [1661412920.284322723] [obstacle_avoidance_planner.time]:     generateFineTrajectoryPoints:= 0.283 [ms]
[component_container_mt-15] [INFO] [1661412920.393833142] [obstacle_avoidance_planner.time]:     generateFineTrajectoryPoints:= 0.227 [ms]
[component_container_mt-15] [INFO] [1661412920.480718442] [obstacle_avoidance_planner.time]:     generateFineTrajectoryPoints:= 0.292 [ms]
[component_container_mt-15] [INFO] [1661412920.583334207] [obstacle_avoidance_planner.time]:     generateFineTrajectoryPoints:= 0.225 [ms]
[component_container_mt-15] [INFO] [1661412920.681510785] [obstacle_avoidance_planner.time]:     generateFineTrajectoryPoints:= 0.27 [ms]
[component_container_mt-15] [INFO] [1661412920.796739779] [obstacle_avoidance_planner.time]:     generateFineTrajectoryPoints:= 0.239 [ms]
[component_container_mt-15] [INFO] [1661412920.897798668] [obstacle_avoidance_planner.time]:     generateFineTrajectoryPoints:= 0.245 [ms]
[component_container_mt-15] [INFO] [1661412920.985440207] [obstacle_avoidance_planner.time]:     generateFineTrajectoryPoints:= 0.281 [ms]
[component_container_mt-15] [INFO] [1661412921.080549138] [obstacle_avoidance_planner.time]:     generateFineTrajectoryPoints:= 0.303 [ms]
[component_container_mt-15] [INFO] [1661412921.241484633] [obstacle_avoidance_planner.time]:     generateFineTrajectoryPoints:= 0.236 [ms]
```

slerp

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [ ] The PR follows the [pull request guidelines].

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

File: common/interpolation/include/interpolation/spline_interpolation.hpp [hunk 1]
```diff
@@ -49,6 +49,9 @@ struct MultiSplineCoef
 std::vector<double> slerp(
   const std::vector<double> & base_keys, const std::vector<double> & base_values,
   const std::vector<double> & query_keys);
+std::array<std::vector<double>, 3> slerp(
+  const std::vector<double> & base_keys, const std::vector<double> & base_x_values,
+  const std::vector<double> & base_y_values, const std::vector<double> & query_keys);
```
Comment (takayuki5168) [inline common/interpolation/include/interpolation/spline_interpolation.hpp:54]: spline_interpolation.hpp is pure spline interpolation implementation.
If you create this function like for point2d, could you move it to spline_interpolation_points_2d.cpp and rename it easier name to understand like slerp2dFromXY (I don't know if this name is easy to understand though).

**Comment by takayuki5168** [comment]
@angry-crab Do you know why there is large calculation cost difference between before and after? Both you set skip_optimization false?

**Comment by angry-crab** [comment]
> @angry-crab Do you know why there is large calculation cost difference between before and after? Both you set skip_optimization ?

It has nothing to do with `skip_optimization`. `fillYawInTrajectoryPoint` performed another interpolation using the previously interpolated result, which contained a large amount of points. The second interpolation was unnecessary.

**Comment by shulanbushangshu** [comment]
@angry-crab @takayuki5168 
Yes,I test the cost of the function "generateFineTrajectoryPoints". 
When the length of path is default value(max:300m),the cost is shown as below:
Screenshot from 2022-08-29 10-33-53
As the comment said,the  calculation is heavy.


The  StepPolygon of obstacle_stop_planner is shown as below:

Screenshot from 2022-08-29 11-12-31

**Comment by shulanbushangshu** [comment]
@angry-crab 
I test the changed code .It works well:
The time cost:
Screenshot from 2022-08-31 15-31-11
The result of path:
Screenshot from 2022-08-31 15-36-44

@takayuki5168 
In the "fillYawInTrajectoryPoint",the most time is cost in the code("slerpYawFromPoints") as below:
  std::vector<double> yaw_vec;
  for (size_t i = 0; i < points.size(); ++i) {
    const double yaw = interpolator.getSplineInterpolatedYaw(i, 0.0);
    yaw_vec.push_back(yaw);
  }
  return yaw_vec;

The code uses too much loop.
In the function"getSplineInterpolatedYaw",it can get all  "diff_x"、"diff_y"in once.But"getSplineInterpolatedYaw" calculates "diff_x\diff_y" one by one.

**Review by takayuki5168 — APPROVED** [review]
LGTM

**Pull request merged**